king_phisher/server/database/models.py

```diff
@@ -397,6 +398,105 @@ class Message(Base):
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	test_answer_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_answers.id'), nullable=False)
+	test_question_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_questions.id'), nullable=False)
+
+@register_table
+class ActualAnswersLinkTestAnswer(Base):
+	__repr_attributes__ = ('test_question_id', 'test_answer_id')
+	__tablename__ = 'actual_answers'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	test_answer_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_answers.id'), nullable=False)
+	test_question_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_questions.id'), nullable=False)
+
+@register_table
+class TestAnswer(Base):
+	__repr_attributes__ = ('answer',)
+	__tablename__ = 'test_answers'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	answer = sqlalchemy.Column(sqlalchemy.String, nullable=False)
+	case_sensitive = sqlalchemy.Column(sqlalchemy.Boolean, default=False)
+	# relationships
+	module_question = sqlalchemy.orm.relationship('TestQuestion', backref='test_answers', cascade='all, delete-orphan')
+	test_submission_link_test_answer = sqlalchemy.orm.relationship('TestSubmissionLinkTestAnswer', backref='test_answers', cascade='all, delete-orphan')
+	test_question_link_test_answer = sqlalchemy.orm.relationship('TestQuestionLinkTestAnswer', backref='test_answers', cascade='all, delete-orphan')
+	actual_answer_link_test_answer = sqlalchemy.orm.relationship('ActualAnswerLinkTestAnswer', backref='test_answers', cascade='all, delete-orphan')
+
+@register_table
+class TestSubmissionLinkTestAnswer(Base):
+	__repr_attributes__ = ('test_submission_id', 'module_question_id', 'answer_id')
+	__tablename__ = 'test_submission_link_test_answer'
+	id = sqlalchemy.Column(sqlalchemy.Integer, primary_key=True)
+	test_submission_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_submissions.id'), nullable=False)
+	test_question_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_questions.id'), nullable=False)
+	test_answer_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('test_answers.id'), nullable=False)
+
+@register_table
+class Test(TagMixIn, Base):
+	__repr_attributes__ = ('name', 'minimum_score')
+	__tablename__ = 'tests'
+	created = sqlalchemy.Column(sqlalchemy.DateTime, default=current_timestamp)
+	minimum_score = sqlalchemy.Column(sqlalchemy.Integer, nullable=False)
+	max_attempts = sqlalchemy.Column(sqlalchemy.Integer, default=1)
+	user_id = sqlalchemy.Column(sqlalchemy.Integer, sqlalchemy.ForeignKey('users.id'), nullable=False)
+	# relationships
+	test_submissions = sqlalchemy.orm.relationship('TestSubmission', backref='tests', cascade='all, delete-orphan')
+	test_link_test_module = sqlalchemy.orm.relationship('TestLinkTestModule', backref='tests', cascade='all, delete-orphan')
```

Comment:
Should there be a `campaigns = sqlalchemy.orm.relationship('Campaign', backref='test', cascade='all, delete-orphan')` here?